### PR TITLE
[JENKINS-73883] Un-inline JavaScript in `history.jelly` and `index.jelly`

### PR DESCRIPTION
**Pull request by shlomomdahan**
[JENKINS-73883](https://issues.jenkins.io/browse/JENKINS-73883)

**HISTORY.JELLY:** 

Removes inline <script> and onClick handler in history.jelly

STATUS_QUO:
https://www.loom.com/share/fb00cb793c5048b8bbadeb0f6607e480

NON_RESTRICTIVE:
https://www.loom.com/share/b11f0f5b2f77479e95aed3eabf58c874

RESTRICTIVE:
https://www.loom.com/share/e5f19e60066b4789b6550d34df470ab0

______________________________________________________

**INDEX.JELLY**

removes inline onClick and <script> tag in index.jelly

STATUS QUO
https://www.loom.com/share/4a7b5d931ce14e018c5a3f1b0056c996?sid=9cae26ed-77f8-4c30-ba43-301f5e0d461e

RESTRICTIVE
https://www.loom.com/share/7201227e5d1b45728630a0426021bc66

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/resources/hudson/plugins/jobConfigHistory/ComputerConfigHistoryAction/index.jelly [hunk 1]
```diff
@@ -9,7 +9,8 @@
       <h1>${%Agent Configuration History}</h1>
       <div>
 
-        <script src="${rootURL}/plugin/jobConfigHistory/deleteRevisionAndTableEntry.js"/>
+        <st:adjunct includes="hudson.plugins.jobConfigHistory.deleteRevisionAndTableEntry" />
```
Comment (basil) [inline src/main/resources/hudson/plugins/jobConfigHistory/ComputerConfigHistoryAction/index.jelly:12]: This is still referenced with a `<script>` tag in `src/main/resources/hudson/plugins/jobConfigHistory/JobConfigHistoryProjectAction/index.jelly`, so that needs to be changed to `<st:adjunct>` for consistency with the other changes in this PR. Once that is done, we can delete `src/main/webapp/deleteRevisionAndTableEntry.js`.

File: src/main/resources/hudson/plugins/jobConfigHistory/ComputerConfigHistoryAction/index.jelly [hunk 2]
```diff
@@ -111,7 +112,7 @@
                         <j:if test="${it.hasDeleteEntryPermission()}">
                           <td style="text-align:center">
                               <j:set var="message" value="${%Do you really want to delete the history entry} "/>
-                              <button type="button" class="jch delete-button" onClick="removeEntryFromTable('table-row-${configNr}', '${config.date}', 'null', '${message}')" value="X">
+                            <button type="button" class="jch delete-button" data-config-nr="${configNr}" data-config-date="${config.date}" data-message-text="${message}" value="X">
```
Comment (basil) [inline src/main/resources/hudson/plugins/jobConfigHistory/ComputerConfigHistoryAction/index.jelly:115]: Unnecessary change in indentation? This makes the diff harder to read.

File: src/main/resources/hudson/plugins/jobConfigHistory/deleteRevisionAndTableEntry.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
+function removeEntryFromTable(id, timestamp, name, message) {
```
Comment (basil) [inline src/main/resources/hudson/plugins/jobConfigHistory/deleteRevisionAndTableEntry.js:1]: Why remove the comment that was present in the original?

**Review by basil — COMMENTED** [review]
Per https://www.jenkins.io/doc/developer/security/csp/#inline-event-handlers

> For event handlers like `onclick` that used to call `return false` to prevent the usual action (e.g. link navigation) from happening, add a call to `Event.preventDefault()` in a separate event handler on the provided `Event` argument.

File: src/main/resources/hudson/plugins/jobConfigHistory/deleteRevisionAndTableEntry.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        xmlHttp.send(null);
+    } else {
+        return false;
+    }
+}
+
+document.addEventListener('DOMContentLoaded', function() {
+
+    const systemConfigDeleteButtons = document.querySelectorAll('.jenkins-button--destructive');
+    const agentConfigDeleteButtons = document.querySelectorAll('.delete-button');
```
Comment (basil) [inline src/main/resources/hudson/plugins/jobConfigHistory/deleteRevisionAndTableEntry.js:25]: Declaration is too far away from its usage, which makes the code harder to read. See e.g. https://checkstyle.sourceforge.io/checks/coding/variabledeclarationusagedistance.html

**Comment by shlomomdahan** [comment]
> Per https://www.jenkins.io/doc/developer/security/csp/#inline-event-handlers
> 
> > For event handlers like `onclick` that used to call `return false` to prevent the usual action (e.g. link navigation) from happening, add a call to `Event.preventDefault()` in a separate event handler on the provided `Event` argument.

@basil this is helpful to note. 

Is any change required in this case though? the onClicks did not make any calls to return false here

**Comment by basil** [comment]
The point here is that there should be a prompt for "Do you really want to delete the history entry", which if the user vetoes (by clicking "no") the entry shouldn't be deleted. I didn't see that in the testing videos, so I wanted to make sure we cover this case.

File: src/main/resources/hudson/plugins/jobConfigHistory/deleteRevisionAndTableEntry.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    agentConfigDeleteButtons.forEach(button => {
+        button.addEventListener('click', function(event) {
+            const configNr = this.getAttribute('data-config-nr');
+            const configDate = this.getAttribute('data-config-date');
+            const message = this.getAttribute('data-message-text');
+            removeEntryFromTable(`table-row-${configNr}`, configDate, null, message);
+        });
+    });
+});
```
Comment (yaroslavafenkin) [inline src/main/resources/hudson/plugins/jobConfigHistory/deleteRevisionAndTableEntry.js:52]: And while doing other changes empty line in the end of the file won't hurt :)

File: src/main/resources/hudson/plugins/jobConfigHistory/deleteRevisionAndTableEntry.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            button.addEventListener('click', function(event) {
+                const configNr = this.getAttribute('data-config-nr');
+                const configDate = this.getAttribute('data-config-date');
+                const message = this.getAttribute('data-message-text');
+                removeEntryFromTable(`table-row-${configNr}`, configDate, jobName, message);
+            });
+        }
+    );
+
+    const agentConfigDeleteButtons = document.querySelectorAll('.delete-button');
```
Comment (yaroslavafenkin) [inline src/main/resources/hudson/plugins/jobConfigHistory/deleteRevisionAndTableEntry.js:42]: Also seems too generic for me.

From https://www.jenkins.io/doc/developer/security/csp/#inline-event-handlers:
> Be mindful of Jenkins’s extensibility, so consider including plugin names in element class names or IDs to prevent unintentional conflicts with other plugins.

**Comment by yaroslavafenkin** [comment]
> The point here is that there should be a prompt for "Do you really want to delete the history entry", which if the user vetoes (by clicking "no") the entry shouldn't be deleted. I didn't see that in the testing videos, so I wanted to make sure we cover this case.

I was confused by that too. I think the prompt was there, just not visible on the recording. Look at the way Shlomo was moving his pointer to the middle of the screen where the prompt usually appears. Definitely would've helped if we saw the prompt too :)

**Comment by yaroslavafenkin** [comment]
> > Per https://www.jenkins.io/doc/developer/security/csp/#inline-event-handlers
> > > For event handlers like `onclick` that used to call `return false` to prevent the usual action (e.g. link navigation) from happening, add a call to `Event.preventDefault()` in a separate event handler on the provided `Event` argument.
> 
> @basil this is helpful to note.
> 
> Is any change required in this case though? the onClicks did not make any calls to return false here

We normally need to `e.preventDefault()` if we click on a button that submits a form. Here there's no form, so that code looks like it does nothing.

**Comment by shlomomdahan** [comment]
> The point here is that there should be a prompt for "Do you really want to delete the history entry", which if the user vetoes (by clicking "no") the entry shouldn't be deleted. I didn't see that in the testing videos, so I wanted to make sure we cover this case.

I didn't realize the prompt window was not visible. You can see below that cancel delete works fine as well.

https://www.loom.com/share/0a0c0cfdab284f33963902fa84b28242?sid=1e2acea6-aa64-4d98-8f01-fe2f2cfa154f

File: src/main/resources/hudson/plugins/jobConfigHistory/deleteRevisionAndTableEntry.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        xmlHttp.setRequestHeader(document.head.getAttribute('data-crumb-header'), document.head.getAttribute('data-crumb-value'));
+        xmlHttp.send(null);
+    } else {
+        return false;
+    }
+}
+
+document.addEventListener('DOMContentLoaded', function() {
+
+    const systemConfigDeleteButtons = document.querySelectorAll('.jenkins-button--destructive');
```
[the remark this answers is not in the document]
Comment (shlomomdahan) [reply src/main/resources/hudson/plugins/jobConfigHistory/deleteRevisionAndTableEntry.js:24]: Good point.. Let me make this change. Thanks @yaroslavafenkin

**Comment by shlomomdahan** [comment]
@yaroslavafenkin 

Thank you for your comments. I've updated the files accordingly.

**Review by yaroslavafenkin — APPROVED** [review]
LGTM

**Review by basil — APPROVED** [review]
Thanks @shlomomdahan! Can you please test this in ATH as in https://github.com/jenkinsci/acceptance-test-harness/pull/1772 and then I will be ready to merge/release this.

**Comment by StefanSpieker** [comment]
Thanks a lot @shlomomdahan and @basil
